go/libraries/doltcore/sqle/dfunctions/dolt_pull.go

```diff
@@ -0,0 +1,200 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dfunctions
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+
+	"github.com/dolthub/dolt/go/cmd/dolt/cli"
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+	"github.com/dolthub/dolt/go/libraries/doltcore/env"
+	"github.com/dolthub/dolt/go/libraries/doltcore/env/actions"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/dsess"
+	"github.com/dolthub/dolt/go/store/datas"
+)
+
+const DoltPullFuncName = "dolt_pull"
+
+type DoltPullFunc struct {
+	expression.NaryExpression
+}
+
+// NewPullFunc creates a new PullFunc expression.
+func NewPullFunc(ctx *sql.Context, args ...sql.Expression) (sql.Expression, error) {
+	return &DoltPullFunc{expression.NaryExpression{ChildExpressions: args}}, nil
+}
+
+func (d DoltPullFunc) String() string {
+	childrenStrings := make([]string, len(d.Children()))
+
+	for i, child := range d.Children() {
+		childrenStrings[i] = child.String()
+	}
+
+	return fmt.Sprintf("DOLT_PULL(%s)", strings.Join(childrenStrings, ","))
+}
+
+func (d DoltPullFunc) Type() sql.Type {
+	return sql.Boolean
+}
+
+func (d DoltPullFunc) WithChildren(ctx *sql.Context, children ...sql.Expression) (sql.Expression, error) {
+	return NewPullFunc(ctx, children...)
+}
+
+func (d DoltPullFunc) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	dbName := ctx.GetCurrentDatabase()
+
+	if len(dbName) == 0 {
+		return noConflicts, fmt.Errorf("empty database name.")
+	}
+
+	sess := dsess.DSessFromSess(ctx.Session)
+	dbData, ok := sess.GetDbData(ctx, dbName)
+
+	if !ok {
+		return noConflicts, fmt.Errorf("Could not load database %s", dbName)
+	}
+
+	ap := cli.CreateMergeArgParser()
+	args, err := getDoltArgs(ctx, row, d.Children())
+
+	apr, err := ap.Parse(args)
+	if err != nil {
+		return noConflicts, err
+	}
+
+	if apr.NArg() > 1 {
+		return noConflicts, actions.ErrInvalidPullArgs
+	}
+
+	var remoteName string
+	if apr.NArg() == 1 {
+		remoteName = apr.Arg(0)
+	}
+
+	pullSpec, err := env.ParsePullSpec(ctx, dbData.Rsr, remoteName, apr.Contains(cli.SquashParam), apr.Contains(cli.NoFFParam), apr.Contains(cli.ForceFlag))
+	if err != nil {
+		return noConflicts, err
+	}
+
+	srcDB, err := pullSpec.Remote.GetRemoteDBWithoutCaching(ctx, dbData.Ddb.ValueReadWriter().Format())
+	if err != nil {
+		return noConflicts, fmt.Errorf("failed to get remote db; %w", err)
+	}
+
+	ws, err := sess.WorkingSet(ctx, dbName)
+	if err != nil {
+		return noConflicts, err
+	}
+
+	var conflicts interface{}
+	for _, refSpec := range pullSpec.RefSpecs {
```

Comment:
Structuring this in a loop is misleading, because the pull logic in the body only happens once (right?)

Use a loop to determine the remoteTrackRef, then outside the loop do a single pull action.

Reply:
i think multiple remote tracking branches per remote is a thing https://git-scm.com/book/en/v2/Git-Internals-The-Refspec